src/cr/cube/slices.py

```diff
@@ -0,0 +1,1140 @@
+
+
+class _CatXCatSlice(object):
+    def __init__(self, dimensions, counts, base_counts):
+        self._dimensions = dimensions
+        self._counts = counts
+        self._base_counts = base_counts
+
+    @lazyproperty
+    def _row_dimension(self):
+        return self._dimensions[0]
+
+    @lazyproperty
+    def _column_dimension(self):
+        return self._dimensions[1]
+
+    @lazyproperty
+    def _row_elements(self):
+        return self._row_dimension.valid_elements
+
+    @lazyproperty
+    def _column_elements(self):
+        return self._column_dimension.valid_elements
+
+    @lazyproperty
+    def _row_generator(self):
+        return zip(self._counts, self._base_counts, self._row_elements)
+
+    @lazyproperty
+    def _column_generator(self):
+        return zip(self._counts.T, self._base_counts.T, self._column_elements)
+
+    @lazyproperty
+    def rows(self):
+        return tuple(
+            _CategoricalVector(counts, base_counts, element.label, self.table_margin)
+            for counts, base_counts, element in self._row_generator
+        )
+
+    @lazyproperty
+    def columns(self):
+        return tuple(
+            _CategoricalVector(counts, base_counts, element.label, self.table_margin)
+            for counts, base_counts, element in self._column_generator
+        )
+
+    @lazyproperty
+    def table_margin(self):
+        return np.sum(self._counts)
+
+
+class _1DCatSlice(_CatXCatSlice):
+    @lazyproperty
+    def columns(self):
+        return tuple(
+            [
+                _CategoricalVector(
+                    self._counts, self._base_counts, "Summary", self.table_margin
+                )
+            ]
+        )
+
+
+class _MrXCatSlice(_CatXCatSlice):
+    @lazyproperty
+    def rows(self):
+        """Use only selected counts."""
+        return tuple(
+            _CategoricalVector(counts[0], base_counts[0], element.label, table_margin)
+            for counts, base_counts, element, table_margin in self._row_generator
+        )
+
+    @lazyproperty
+    def _row_generator(self):
+        return zip(
+            self._counts, self._base_counts, self._row_elements, self.table_margin
+        )
+
+    @lazyproperty
+    def columns(self):
+        """Use bother selected and not-selected counts."""
+        return tuple(
+            _MultipleResponseVector(
+                counts, base_counts, element.label, self.table_margin
+            )
+            for counts, base_counts, element in self._column_generator
+        )
+
+    @lazyproperty
+    def table_margin(self):
+        # return np.sum(np.array([col.margin for col in self.columns]), axis=0)[:, None]
+        return np.sum(self._counts, axis=(1, 2))
+
+
+class _1DMrSlice(_MrXCatSlice):
+    @lazyproperty
+    def columns(self):
+        return tuple(
+            [
+                _MultipleResponseVector(
+                    self._counts.T, self._base_counts.T, "Summary", self.table_margin
```

Comment:
Label says "Summary", cause why not :) I'm open for better ideas. It's the name of the opposite dimension labels, which don't exist for 1D vector.